### PR TITLE
Make image archive/save idempotent, using image id and repo tags as keys

### PR DESCRIPTION
**Pull request by iamjpotts**
##### SUMMARY

The `docker_image` component supports saving images to a tar file, similar to `docker image save` command. When a `docker_image` archive task is run repeatedly, it recreates the file every time.

After this change, the task will compare the Docker image ID of the archive file (if any) vs the image stored in the Docker Engine, and if they are the same, the archive file is left untouched.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

`docker_image` used with `archive_path`

##### ADDITIONAL INFORMATION

Consider a scenario with two servers (`localhost` and `target`), where an Ansible playbook synchronizes images to a target server.

1. Playbook run against `localhost` uses `docker_load` with `archive_path` to export an image to a tar
2. Playbook run using `ansible.builtin.file` synchronizes the directory with the tar file on localhost with a target server directory
3. Playbook run using `docker_load` imports tar file on target server to target server's docker engine

Currently, this workflow regenerates the tar files over and over again, resending them to the target server each time.

After this change, Step 1 only exports the tar files once, until the source image changes, preventing the network copy from `localhost` to `target` on subsequent runs.
 
Export playbook

```
- hosts: localhost
  connection: local
  vars:
    docker_images:
      - nginx:1.20

  tasks:

    - name: Docker image export directory 
      ansible.builtin.file:
        path: ~/.docker-image-cache
        state: directory
        mode: '0755'

    - name: Export Docker Images
      community.docker.docker_image:
        name: nginx:1.20
        source: local
        archive_path: ~/.docker-image-cache/nginx-1.20.tar
        state: present
```

Import playbook

```

- hosts: target-server
  tasks:

    - name: Sync cache directory
      ansible.builtin.copy:
        src: ~/.docker-image-cache
        dest: ~/

    - name: Import Docker Images
      community.docker.docker_image:
        name: nginx:1.20
        source: load
        load_path: ~/.docker-image-cache/nginx-1.20.tar
        state: present

```

Output, without this PR, run 1 of 2.

```
PLAY [localhost] **************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Docker image export directory] ******************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Export Docker Images] ***************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP ********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

Output, without this PR, run 2 of 2. Shows `changed` again

```

PLAY [localhost] **************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Docker image export directory] ******************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Export Docker Images] ***************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP ********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

Output, with this PR, run 1 of 2.

```
PLAY [localhost] **************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Docker image export directory] ******************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Export Docker Images] ***************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP ********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

Output, with this PR, run 2 of 2. Shows `ok` on second run.

```

PLAY [localhost] **************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Docker image export directory] ******************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Export Docker Images] ***************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

PLAY RECAP ********************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

### CODE REVIEW

**Comment by iamjpotts** [comment]
Had some difficulty running `ansible-test` locally; would like a maintainer to approve running the CI so I can compare its output vs local.

**Comment by iamjpotts** [comment]
@felixfontein 

I've removed the 'do not merge' label.

I attempted to add two `.tar` files for use with unit tests but they got flagged by the license checker, since they are not text files and can't be checked for a copyright header.

I worked around that by base64-encoding the files and including them as a string literal.

The files were generated by me using `docker image save` - is this approach suitable for the licensing issue, or should another be considered?

**Comment by felixfontein** [comment]
> I attempted to add two `.tar` files for use with unit tests but they got flagged by the license checker, since they are not text files and can't be checked for a copyright header.
> 
> I worked around that by base64-encoding the files and including them as a string literal.
> 
> The files were generated by me using `docker image save` - is this approach suitable for the licensing issue, or should another be considered?

I would really avoid adding binary blobs to this repository. How about creating these files during the integration tests (by running `docker image save`) and then using the freshly generated files?

**Comment by iamjpotts** [comment]
One of the two tar files is very small (hello world is about 3kb). The other one is larger (about 3 mb) but I can find a smaller substitute, for example creating an image from `scratch` and adding a text file to it (e.g a copyright notice).

Would two small files be adequate? Is there a way to exclude them from release/distribution packaging?

The unit tests are easier and faster to run, and in principle, I think these tests belong there. They test parsing a file, as opposed to interacting with a service.

Regenerating them each time also introduces variability into the test input; it's nice to have the same exact input across all versions of Python being tested.

**Comment by felixfontein** [comment]
> Would two small files be adequate? Is there a way to exclude them from release/distribution packaging?

I would **really** avoid adding any kind of binary file, if not absolutely necessary.

> The unit tests are easier and faster to run, and in principle, I think these tests belong there. They test parsing a file, as opposed to interacting with a service.

They are faster to run, but they also are a big problem, since you are including copyrighted contents into this repository. Both the hello-world executable and busybox come with copyright and legal conditions attached.

> Regenerating them each time also introduces variability into the test input; it's nice to have the same exact input across all versions of Python being tested.

I don't think this will be a problem.

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:933]: Better use `PY2` (`from ansible.module_utils.six import PY2`) to handle these differences.

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:971]: Same here.

File: .gitignore
Comment (iamjpotts) [reply .gitignore:139]: Done, see #501

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py:591]: Earlier in the function, there's already a guard on `image` with a `return` to exit the function.

https://github.com/ansible-collections/community.docker/blob/main/plugins/modules/docker_image.py#L548

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py:590]: I'll delete it to simplify the PR.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py:560]: OK.

What are the differences between `self.log` and `module.debug` ?

I have no opinion either way, but want to understand them better.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py:960]: Ugh, my exception handling looks awful, I should have stopped before 4 am.

**Comment by iamjpotts** [comment]
Changes:

* Added change log fragment
* Replaced committed base-64 encoded tar files with runtime-generated stubs with same structure
* Fixed nested exception handling and added tests for causes
* Replaced action entries with `log.debug` statements (`self.log` calls an empty function with `pass` inside)
* Replaced `__exit__` attribute conditionals with `close()` calls
* Removed `old_image_id` result

File: changelogs/fragments/500-idempotent-image-archival.yaml
Comment (felixfontein) [inline changelogs/fragments/500-idempotent-image-archival.yaml:2]:
```suggestion
  - docker_image - when using ``archive_path``, detect whether changes are necessary based on the image ID (hash). If the existing tar archive matches the source, do nothing.
```

File: changelogs/fragments/500-idempotent-image-archival.yaml
Comment (felixfontein) [inline changelogs/fragments/500-idempotent-image-archival.yaml:3]:
```suggestion
    Previously, each task execution re-created the archive
    (https://github.com/ansible-collections/community.docker/pull/500).
```

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:378]: Please remove this.

File: plugins/modules/docker_image.py
Comment (felixfontein) [reply plugins/modules/docker_image.py]: `self.log()` is the preferred way for the Docker modules, since all other Docker modules also use this. It does nothing, but can easily be enable to print something when debugging a module in the collection.

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:567]: Please use `self.log()` or - if really necessary - `module.debug()`.

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:576]: This only needs to be added if `changed` is `True`. Otherwise, the module is not doing something, and adding this action would then be wrong.

**Comment by felixfontein** [comment]
You have a very strange commit list, I guess you should probably rebase to get rid of all the stray changes which do not belong into this PR.

File: plugins/module_utils/image_archive.py
Comment (felixfontein) [inline plugins/module_utils/image_archive.py:33]: I don't think this function belongs in this file. It should be part of the docker_image module.

**Comment by felixfontein** [comment]
PS: The `[DO NOT MERGE]` prefix is for PRs that will never be merged and are only for testing something. I've marked your PR as a draft, if you want to use a prefix use `[WIP]` or `WIP:`.

File: plugins/module_utils/image_archive.py
Comment (iamjpotts) [reply plugins/module_utils/image_archive.py:33]: Heh I thought about that and waffled on it.

You prefer a standalone function or a method?

**Comment by iamjpotts** [comment]
> You have a very strange commit list, I guess you should probably rebase to get rid of all the stray changes which do not belong into this PR.

Yeah it's screwed up bc of when I rebased vs when the source main `branch` was updated yesterday. I'll rebase again soon.

File: plugins/module_utils/image_archive.py
Comment (felixfontein) [reply plugins/module_utils/image_archive.py:33]: Considering that most functionality in that module is a method, I would make it a method there.

Comment (iamjpotts) [reply plugins/module_utils/image_archive.py:33]: There's some shared functions for stubbing tar files that would be needed by both tests in `modules/` and `module_utils/` and I could create a conftest.py in their parent to share them, but that scopes them across all tests, as opposed to scoping them to just `docker_image`.

Is that OK?

If I leave `def archived_image_action` where it is, the new `conftest.py` isn't needed at all.

Comment (felixfontein) [reply plugins/module_utils/image_archive.py:33]: You can also put unit test code in arbitrary `.py` files and just import it. There's no need to always use `conftest.py` for that.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: [This line](https://github.com/ansible-collections/community.docker/pull/500/files#diff-6178e5dc28781ab3227b344e50384dafb10d5df4aa0b54dc6fa3585c5c259d14R549) is the only thing logged now:

It logs an exception, if any, from reading or parsing the tar file:

```
'Unable to extract manifest summary from archive: %s' % to_native(exc)
```

How do I call `module.debug` ?

That doesn't resolve in my IDE.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: Removed.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: Changed to self.log, but would like to know how to call `module.debug` to report the exception.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: Done.

File: plugins/module_utils/image_archive.py
Comment (iamjpotts) [reply plugins/module_utils/image_archive.py]: Added a new file `tests/unit/plugins/test_support/docker_image_archive_stubbing.py`

**Comment by iamjpotts** [comment]
I refactored `archived_image_action` onto `ImageManager` but stopped short of making it an instance method due to all the bootstrapping / mocking that would be required.

`ImageManager` has several "sub tasks" - archival being one of them - that would be easier to test if they were composed from separate classes, but that would be way too invasive of a refactor here.

File: tests/integration/targets/docker_image/tasks/tests/options.yml
Comment (felixfontein) [inline tests/integration/targets/docker_image/tasks/tests/options.yml:8]:
```suggestion
```
`vars/main.yml` is automatically loaded and made available in a role.

File: tests/integration/targets/docker_image/tasks/tests/options.yml
Comment (felixfontein) [inline tests/integration/targets/docker_image/tasks/tests/options.yml:282]: You need to register this image name in line 16 (`inames: "{{ inames + [iname, iname_1, docker_test_image_hello_world_archive_alt] }}"`).

File: tests/integration/targets/docker_image/vars/main.yml
Comment (felixfontein) [inline tests/integration/targets/docker_image/vars/main.yml:6]: Why use a fixed name and not generate one (like `iname` and `iname_1`)? That way the name should be unique per test, and potential leftovers from previous runs do not cause problems.

File: tests/unit/plugins/modules/test_docker_image.py
Comment (felixfontein) [inline tests/unit/plugins/modules/test_docker_image.py:27]: The name of this function is very generic; it should be more specific since this function tests for a very specific input.

File: tests/unit/plugins/test_support/docker_image_archive_stubbing.py
Comment (felixfontein) [inline tests/unit/plugins/test_support/docker_image_archive_stubbing.py:41]: Shouldn't the `try:` start here, since the temp file has already been created and can be removed if the following write fails for whatever reason?

File: tests/unit/plugins/test_support/docker_image_archive_stubbing.py
Comment (felixfontein) [inline tests/unit/plugins/test_support/docker_image_archive_stubbing.py:65]: Same here.

File: tests/integration/targets/docker_image/tasks/tests/options.yml
Comment (iamjpotts) [reply tests/integration/targets/docker_image/tasks/tests/options.yml:282]: Why did the tests work without doing this, and what's gained by adding it here?

Is this required due to removing the new `include_vars` task from the top of `options.yml` ?

File: tests/integration/targets/docker_image/vars/main.yml
Comment (iamjpotts) [reply tests/integration/targets/docker_image/vars/main.yml:6]: The other examples I saw used fixed names, e.g. `tests/integration/targets/setup_docker_registry/vars/main.yml` (granted, those are referencing external images).

How do I auto-generate a name?

File: tests/unit/plugins/modules/test_docker_image.py
Comment (iamjpotts) [reply tests/unit/plugins/modules/test_docker_image.py:27]: Ironically that function is not used after the refactoring so I'm deleting it.

File: tests/unit/plugins/test_support/docker_image_archive_stubbing.py
Comment (iamjpotts) [reply tests/unit/plugins/test_support/docker_image_archive_stubbing.py:41]: Yes, good catch.

I also moved the inner file management inside of the tar file `try..finally`

File: tests/unit/plugins/test_support/docker_image_archive_stubbing.py
Comment (iamjpotts) [reply tests/unit/plugins/test_support/docker_image_archive_stubbing.py]: Another great catch, made same two changes (your suggestion + moving inner file create/delete inside of tar file try..finally)

File: tests/integration/targets/docker_image/tasks/tests/options.yml
Comment (iamjpotts) [reply tests/integration/targets/docker_image/tasks/tests/options.yml:279]: nvm answered my own question by trying it lol

File: tests/integration/targets/docker_image/tasks/tests/options.yml
Comment (iamjpotts) [reply tests/integration/targets/docker_image/tasks/tests/options.yml]: Done.

File: plugins/modules/docker_image.py
Comment (felixfontein) [reply plugins/modules/docker_image.py]: Did you look in https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1305 ?

File: tests/integration/targets/docker_image/vars/main.yml
Comment (felixfontein) [reply tests/integration/targets/docker_image/vars/main.yml:6]: Check out lines 8 and 9 in that file. If you want extra randomness add something like `(10000 | random)`.

Comment (iamjpotts) [reply tests/integration/targets/docker_image/vars/main.yml]: I moved it. I had to add an explict tag suffix because some tasks were interpreting it using `:latest` and others interpreted it using `:hello-world`

**Comment by iamjpotts** [comment]
Made one more docstring change to tickle the build after a transient network related failure.

Ready for review.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: Reverted.

There's problems with the imports that using a relative import fixes.

I'll open a separate issue.

File: plugins/plugin_utils/common_api.py
Comment (iamjpotts) [reply plugins/plugin_utils/common_api.py]: Done.

File: plugins/modules/docker_image.py
Comment (iamjpotts) [reply plugins/modules/docker_image.py]: #507

File: plugins/modules/docker_image.py
Comment (felixfontein) [inline plugins/modules/docker_image.py:346]:
```suggestion
from ansible_collections.community.docker.plugins.module_utils.image_archive import (
```

**Comment by iamjpotts** [comment]
I'm now getting build failures which are timeouts waiting for the temporary docker registry to start.

[For example](https://dev.azure.com/ansible/community.docker/_build/results?buildId=61726&view=logs&j=7eeace52-ac01-56c9-ddb0-19dd66139108&t=2915ff0d-5d52-54f6-b188-708c94614a79)

Made a whitespace change to tickle the build and it failed the same way (three times so far).

I don't _think_ this is related to accepting your import change suggestion, but I rebased right away.

Looking at the CI log, I suspect the registry container is failing to stay running (the previous successful task is a start request), but there isn't any log output from the registry container in the CI log to see why.

**Comment by felixfontein** [comment]
This might be caused by https://github.com/ansible/ansible/pull/78550. Simply ignore that failure for now.

**Comment by felixfontein** [comment]
Now there's a conflict due to the other PR being merged. After rebasing with the latest `main` branch all tests should also pass.

**Review by felixfontein — APPROVED** [review]

**Comment by iamjpotts** [comment]
> Now there's a conflict due to the other PR being merged. After rebasing with the latest `main` branch all tests should also pass.

I'm updating the docstring style in this PR also, it's not committed yet. I'll drop a message here once the build is green and I've looked over the code one last time.

**Review by felixfontein — APPROVED** [review]

**Comment by iamjpotts** [comment]
Build is green.

I did miss one triple-quote to single quote in a [test_docker_image.py](https://github.com/ansible-collections/community.docker/pull/500/files#diff-f918e6be7fd2af82f28f70c43902120ed64d6460ae46c8a6957b9a8ad5e67f83R35)

If that doesn't concern you, I don't see anything else to change.

**Comment by felixfontein** [comment]
@iamjpotts thanks a lot for your contribution!

**Pull request merged**
